### PR TITLE
wait-online: add --any option

### LINKED ISSUE #9714: systemd-networkd-wait-online should have an option to wait until any 1 of multiple interfaces are online.
systemd-networkd-wait-online.service has an option to wait until 1 or more specified interfaces are configured, or by default waits until all interfaces it manages are configured.

It does not have an option to wait until any one of all (or specified) interfaces are configured.

This would be useful for servers that have multiple ports, all are left to auto-configure, but only one will ever be plugged in.

man page is here showing this limitation:
https://www.freedesktop.org/software/systemd/man/systemd-networkd-wait-online.service.html

Perhaps add an option:
--wait-for-first-interface

### PR DESCRIPTION
**Pull request by yuwata**
When this option is specified, wait-online exits with success even when several interfaces are in configuring state.

Closes #9714.

### CODE REVIEW

**Comment by yuwata** [comment]
Maybe '--any' is better for the name of the new option. C.f. https://github.com/systemd/systemd/issues/5154#issuecomment-312845422.

**Comment by yuwata** [comment]
Updated, and renamed to '--any'. I will try to add a testcase.

**Comment by yuwata** [comment]
A test for the new option is added.

File: man/systemd-networkd-wait-online.service.xml [hunk 1]
```diff
@@ -85,6 +85,17 @@
         </para></listitem>
       </varlistentry>
 
+      <varlistentry>
+        <term><option>--any</option></term>
+
+        <listitem><para>Even if several interfaces are in configuring state,
+        <command>systemd-networkd-wait-online</command> exits with success when at least one interface
+        becomes online. When this option is specified with <option>--interface</option>, then
```
Comment (poettering) [inline man/systemd-networkd-wait-online.service.xml]: following our usual syntax please suffix `--interface` with `=` to indicate it expects an argument. i.e. `<option>--interface</option>` → `<option>--interface=</option>`

File: src/network/wait-online/wait-online.c [hunk 2]
```diff
@@ -41,6 +42,7 @@ static int help(void) {
                "     --ignore=INTERFACE     Don't take these interfaces into account\n"
                "  -o --operational-state=OPERSTATE\n"
                "                            Required operational state\n"
+               "     --any                  Wait for until one of the interfaces to be online\n"
```
Comment (poettering) [inline src/network/wait-online/wait-online.c]: Wait for until → Wait until at least
to be online → is online

i.e. `Wait until at leaset one of the interfaces is online`

**Comment by yuwata** [comment]
@poettering Thank you for the comments. Updated.

**Comment by anjalidk** [comment]
Hi, which release is this fix a part of? How can I trace that?

**Comment by anjalidk** [comment]
I mean how can I know which RH systemd version includes this fix?

**Comment by yuwata** [comment]
@anjalidk The option was introduced since v242. (I heard that RHEL8 does not have networkd. Please try to use EPEL package. EPEL has systemd-networkd from v246.)

**Comment by anjalidk** [comment]
@yuwata In my case, I am using OL8 which is based on RHEL8 and I get this failure:
[root@ol8u3-test tmp]# systemctl status NetworkManager-wait-online.service
● NetworkManager-wait-online.service - Network Manager Wait Online
   Loaded: loaded (/usr/lib/systemd/system/NetworkManager-wait-online.service; >
   Active: failed (Result: exit-code) since Thu 2020-11-26 17:12:56 GMT; 20min >
     Docs: man:nm-online(1)
 Main PID: 3640 (code=exited, status=1/FAILURE)

This fails - is there any way I can use this --any option in above network online wait code? I ported your fix in the commit above to OL8 (without the networkd tests) - and want to test it now - but how to enable --any? Since we are using network manager wait online (which is same code that you changed), we can make use of --any here too?

**Comment by anjalidk** [comment]
This is the wait online service file:
[Unit]
Description=Network Manager Wait Online
Documentation=man:nm-online(1)
Requires=NetworkManager.service
After=NetworkManager.service
Before=network-online.target

[Service]
# `nm-online -s` waits until the point when NetworkManager logs
# "startup complete". That is when startup actions are settled and
# devices and profiles reached a conclusive activated or deactivated
# state. It depends on which profiles are configured to autoconnect and
# also depends on profile settings like ipv4.may-fail/ipv6.may-fail,
# which affect when a profile is considered fully activated.
# Check NetworkManager logs to find out why wait-online takes a certain
# time.

Type=oneshot
ExecStart=/usr/bin/nm-online -s -q
RemainAfterExit=yes

# Set $NM_ONLINE_TIMEOUT variable for timeout in seconds.
# Edit with `systemctl edit NetworkManager-wait-online`.
#
# Note, this timeout should commonly not be reached. If your boot
# gets delayed too long, then the solution is usually not to decrease
# the timeout, but to fix your setup so that the connected state
# gets reached earlier.
Environment=NM_ONLINE_TIMEOUT=30

[Install]

Can the --any be added here?

**Comment by yuwata** [comment]
@anjalidk This PR is for systemd-networkd-wait-online.service. Please contact to NetworkManager team if you use NetworkManager-wait-online.service.

**Comment by anjalidk** [comment]
@yuwata thanks, can you point me to the github page for team working on NetworkManager-wait-online.service?

**Comment by yuwata** [comment]
They seem to use gitlab: https://gitlab.freedesktop.org/NetworkManager/NetworkManager
And wait-online source seems https://gitlab.freedesktop.org/NetworkManager/NetworkManager/-/blob/master/clients/nm-online.c
Also, there is a mirror in github: https://github.com/NetworkManager/NetworkManager

**Pull request merged**